cmd/tusd/cli/hooks/proto/v1/hook.proto

```diff
@@ -0,0 +1,66 @@
+syntax = "proto3";
+package v1;
+
+// Uploaded data
+message Upload {
+	// Unique integer identifier of the uploaded file
+	string id = 1;
+	// Total file size in bytes specified in the NewUpload call
+	int64 Size = 2; 
+	// Indicates whether the total file size is deferred until later
+	bool SizeIsDeferred = 3;
+	// Offset in bytes (zero-based)
+	int64 Offset = 4;
+	map<string, string> metaData = 5;
+	// Indicates that this is a partial upload which will later be used to form
+	// a final upload by concatenation. Partial uploads should not be processed
+	// when they are finished since they are only incomplete chunks of files.
+	bool isPartial = 6; 
+	// Indicates that this is a final upload
+	bool isFinal = 7; 
+	// If the upload is a final one (see IsFinal) this will be a non-empty
+	// ordered slice containing the ids of the uploads of which the final upload
+	// will consist after concatenation.
+	repeated string partialUploads = 8;
+	// Storage contains information about where the data storage saves the upload,
+	// for example a file path. The available values vary depending on what data
+	// store is used. This map may also be nil.
+	map <string, string> storage = 9;
+}
+
+message HTTPRequest {
+	// Method is the HTTP method, e.g. POST or PATCH
+	string method = 1;
+	// URI is the full HTTP request URI, e.g. /files/fooo
+	string uri = 2;
+	// RemoteAddr contains the network address that sent the request
+	string remoteAddr = 3; 
+}
+
+// Hook's data
+message Hook {
+	// Upload contains information about the upload that caused this hook
+	// to be fired.
+	Upload upload = 1;
+	// HTTPRequest contains details about the HTTP request that reached
+	// tusd.
+	HTTPRequest httpRequest = 2;
+}
+
+// Request data to send hook
+message SendRequest {
+  string Api = 1;
+  Hook hook = 2;
+}
+
+// Response that contains data for sended hook
+message SendResponse {
+  string Api = 1;
+  int64 statusCode = 2;
```

Comment:
HTTP and file hooks can also return data which is then sent to the client as the HTTP response body if the hook errors out. This feature is still missing here, isn't it?

Reply:
What kind of response would be here?
Just an error string? Or any kind of response? If we want a general struct, we can use [google.protobuf.Any](https://developers.google.com/protocol-buffers/docs/proto3#any) type.

Reply:
The response is a byte array. Since those responses are usually rather small, I think it's fine to transfer them in one go without streaming or anything.

For reference, here is the code for getting the response body for the HTTP hooks: https://github.com/tus/tusd/blob/master/cmd/tusd/cli/hooks/http.go#L55

Reply:
> HTTP and file hooks can also return data which is then sent to the client as the HTTP response body if the hook errors out.

I don't understand why we need to return data if an error occurs. Any error message should go inside error parameter. In the HTTP hook no data is returned when the request gets an error for example:
https://github.com/tus/tusd/blob/master/cmd/tusd/cli/hooks/http.go#L51

My proposal is to leave just a response in the `SendResponse` message and get [Status](https://godoc.org/google.golang.org/grpc/status) data from the error

```
message SendResponse {
  google.protobuf.Any response = 1;
}

WDYT?
```

Reply:
What I basically want is that a hook should be able to provide an error message which is then sent to the client, if the hook fails. This is used for displaying a reason to the end user if the pre-create hook rejects an upload.

I haven't fully understood gRPC's error model, so maybe that is already possible with your PR and we don't need any modifications here.

Reply:
From tusd server we'are making a request to a hook service like this:

```
func (g GrpcHook) InvokeHook(typ HookType, info handler.HookEvent, captureOutput bool) ([]byte, int, error) {
  ...
  resp, err := g.Client.Send(ctx, req)
  ...
}
```

The hook service will have a service that implements `Send` method like this:
```
func (s *Service) Send(ctx context.Context, req *SendRequest) (*SendResponse, error) {
  ...
  return resp, error
}
```

If an error occurs, the hook service could send it back, and the same error is returned to the end user.

Reply:
Ok, then let's leave it like that.